Revise arm platform notes regarding soft float

thejpster opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

This PR updates the Arm microcontroller platform docs to recommend `-fpregs` instead of `+soft-float` as [discussed on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.60-Ctarget-feature.3D.2Bsoft-float.60.20considered.20harmful)

thejpster committed:
Fix typos in arm-none-eabi.md

These arrived in #125690, I think by mistake.

Urgau commented on src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md:
```diff
@@ -50,6 +50,9 @@ to use these flags.
 | Cortex-M7F | SP  | Yes | `cortex-m7` | `-fp64`         |
 | Cortex-M7F | DP  | Yes | `cortex-m7` | None            |
 
+**Note:** Never use the `-fpregs` *target-feature* with these `eabihf` targets
+as it will cause compilation units to have different ABIs, which is unsound.
```
```suggestion
<div class="warning">

Never use the `-fpregs` *target-feature* with these `eabihf` targets as it
will cause compilation units to have different ABIs, which is unsound.

</div>
```
cf. https://rust-lang.github.io/mdBook/format/mdbook.html#classwarning

Urgau commented on src/doc/rustc/src/platform-support/thumbv8m.main-none-eabi.md:
```diff
@@ -74,6 +74,9 @@ to use these flags.
 | Cortex-M85  | DP  | Yes | Int       | `cortex-m85`  | `-mve.fp`             |
 | Cortex-M85  | DP  | Yes | Int+Float | `cortex-m85`  | None                  |
 
+**Note:** Never use the `-fpregs` *target-feature* with these `eabihf` targets
+as it will cause compilation units to have different ABIs, which is unsound.
```
```suggestion
<div class="warning">

Never use the `-fpregs` *target-feature* with these `eabihf` targets as it
will cause compilation units to have different ABIs, which is unsound.

</div>
```

thejpster committed:
Suggested -fpregs instead of +soft-float.

Has the same effect, but turning off a feature matches the other flags better than turning on a feature (which actually turns off a feature).

thejpster committed:
Add warning to arm-none-eabi.md instead of a Note:

ehuss commented:
@rustbot ping arm

Can an Arm maintainer review this, please?

Or, cc @RalfJung, since I've noticed the two of you discussing this on Zulip, if you want to review this.

thejpster commented:
AIUI there are no Arm maintainers for M-Profile, only for the Aarch A-Profile. Although they offered to help on a best-effort basis.

jamesmunns commented:
CC @adamgrieg @thalesfragoso @newAM

jacobbramley commented on src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md:
```diff
@@ -50,6 +50,13 @@ to use these flags.
 
+<div class="warning">
+
+Never use the `-fpregs` *target-feature* with these `eabihf` targets
```
What does "these" refer to? Should it just be "... with `eabihf` targets ...", or perhaps "... with hard-float ABI targets (such as `eabihf`) ..." ?

thejpster replied:
It should say 'this target'. Broader words can go in the arm target docs.

thejpster replied:
Fixed

jacobbramley reviewed (approved):
This seems to do the right thing (at least for the cases I tried and based on the Zulip discussion), and no alarm bells went off when I discussed it with our LLVM teams.

One caveat we noticed — which is really a separate concern — is that there's no way to specify eabihf with no FP. That sounds nonsensical, but it's potentially useful with integer MVE, which uses the same registers. Clang [achieves that with `+nofp`](https://github.com/llvm/llvm-project/blob/55d51dd9dca8220ffaf9260d56dae9f5c34b7120/clang/lib/Driver/ToolChains/Arch/ARM.cpp#L701), which seems to get [translated into `-mfpu=none`](https://github.com/llvm/llvm-project/blob/55d51dd9dca8220ffaf9260d56dae9f5c34b7120/clang/lib/Driver/ToolChain.cpp#L259), but I can't find a way to control that through Rust.

thejpster commented:
Does Integer MVE let you use s0 and d0 even without an FPU?! Wild.

jacobbramley commented:
> Does Integer MVE let you use s0 and d0 even without an FPU?! Wild.

Yes, at least for the few instructions you'd need to use this ABI. I've not done any real work on these devices so can't claim to be an expert, but the [Armv8-M reference manual](https://developer.arm.com/documentation/ddi0553/latest/) says that the necessary `VMOV` instructions are available with either FP or MVE.

The FP/vector register file is overlaid so that `q0` aliases `d1:d0` and `s3:s2:s1:s0`, just like A-class AArch32's Neon registers. You could read `d0` as "the lower 64 bits of `q0`", which might feel more natural in a world without FP hardware. There's a layout diagram under rule R<sub>BBYH</sub>, page 180 in [version DDI0553B.y](https://developer.arm.com/documentation/ddi0553/by/?lang=en) of the manual.

thejpster committed:
Clarify wording around use of `-fpregs` for amr platform docs.

thejpster committed:
Note Integer MVE hard-float use-case in arm platform docs.

thejpster commented:
> > Does Integer MVE let you use s0 and d0 even without an FPU?! Wild.
> 
> Yes, at least for the few instructions you'd need to use this ABI. I've not done any real work on these devices so can't claim to be an expert, but the [Armv8-M reference manual](https://developer.arm.com/documentation/ddi0553/latest/) says that the necessary `VMOV` instructions are available with either FP or MVE.
> 
> The FP/vector register file is overlaid so that `q0` aliases `d1:d0` and `s3:s2:s1:s0`, just like A-class AArch32's Neon registers. You could read `d0` as "the lower 64 bits of `q0`", which might feel more natural in a world without FP hardware. There's a layout diagram under rule RBBYH, page 180 in [version B.y](https://developer.arm.com/documentation/ddi0553/by/?lang=en) of the manual.

OK, I added some notes to mention that this is a potentially valid setup.

ehuss commented:
Thanks! If anyone has any objections or changes, feel free to post an update.

@bors r+ rollup

bors commented:
:pushpin: Commit 5cc1c7b594d97245d2feffbe14478b15f4e48db8 has been approved by `ehuss`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ehuss 5cc1c7b594d97245d2feffbe14478b15f4e48db8 -->
<!-- homu: {"type":"Approved","sha":"5cc1c7b594d97245d2feffbe14478b15f4e48db8","approver":"ehuss","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5cc1c7b594d97245d2feffbe14478b15f4e48db8 with merge edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974...
<!-- homu: {"type":"BuildStarted","head_sha":"5cc1c7b594d97245d2feffbe14478b15f4e48db8","merge_sha":"edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11442411546/job/31842614679)
Approved by: ehuss
Pushing edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11442411546/job/31842614679"},"merge_sha":"edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974"} -->

rust-timer commented:
Finished benchmarking commit (edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974): [comparison URL](https://perf.rust-lang.org/compare.html?start=3ec4308f6cb4bba3140d736d9ebd53b234fa7df8&end=edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=3ec4308f6cb4bba3140d736d9ebd53b234fa7df8&end=edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.9% | [2.9%, 2.9%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.9% | [2.9%, 2.9%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3ec4308f6cb4bba3140d736d9ebd53b234fa7df8&end=edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974&stat=max-rss)
<details>
<summary>Results (primary 2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.3% | [2.3%, 2.3%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.3% | [2.3%, 2.3%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3ec4308f6cb4bba3140d736d9ebd53b234fa7df8&end=edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974&stat=cycles:u)
<details>
<summary>Results (primary 2.8%, secondary -7.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.8%, 2.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -7.7% | [-8.4%, -7.3%] | 3     |
| All ❌✅ (primary)                 | 2.8%  | [2.8%, 2.8%]   | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3ec4308f6cb4bba3140d736d9ebd53b234fa7df8&end=edbd9398dcc0e3b6ffe0dd0cd4578d37d1d36974&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 781.937s -> 781.116s (-0.10%)
**Artifact size**: 333.66 MiB -> 333.71 MiB (0.02%)

Kobzol commented:
Noise, this only updated documentation.

@rustbot label: +perf-regression-triaged
